# Matrix conditions

In [16]:
import numpy as np

In [ ]:
# Well-conditioned system
A1 = np.array([[1, 2],
               [1, 1]])
b1 = np.array([3, 2])

# Solve the system
x1 = np.linalg.solve(A1, b1)
print("Well-conditioned system solution (exact):", x1)

# Slight perturbation in b
b1_perturbed = np.array([3.00001, 2])
x1_perturbed = np.linalg.solve(A1, b1_perturbed)
print("Well-conditioned system solution (perturbed):", x1_perturbed)

# Ill-conditioned system
A2 = np.array([[1.00001, 1],
               [1,       1]])
b2 = np.array([2.00001, 2])

# Solve the system
x2 = np.linalg.solve(A2, b2)
print("Ill-conditioned system solution (exact):", x2)

# Slight perturbation in b
b2_perturbed = np.array([2, 2])
x2_perturbed = np.linalg.solve(A2, b2_perturbed)
print("Ill-conditioned system solution (perturbed):", x2_perturbed)


Well-conditioned system solution (exact): [1. 1.]
Well-conditioned system solution (perturbed): [0.99999 1.00001]
Ill-conditioned system solution (exact): [1. 1.]
Ill-conditioned system solution (perturbed): [0. 2.]


In [5]:
import numpy as np

cond_A1 = np.linalg.cond(A1)
cond_A2 = np.linalg.cond(A2)

print("Condition number of A1 (well-conditioned):", cond_A1)
print("Condition number of A2 (ill-conditioned):", cond_A2)

Condition number of A1 (well-conditioned): 6.854101966249681
Condition number of A2 (ill-conditioned): 400002.0000100217


Okay I checked the condition your way and got
Condition number of A1 (well-conditioned): 6.854101966249681
Condition number of A2 (ill-conditioned): 400002.0000100217

And then I did:
import numpy as np

cond_A1 = np.linalg.norm(A1,np.inf)
cond_A2 = np.linalg.norm(A2,np.inf)

print("Condition number of A1 (well-conditioned):", cond_A1)
print("Condition number of A2 (ill-conditioned):", cond_A2)

Condition number of A1 (well-conditioned): 3.0
Condition number of A2 (ill-conditioned): 2.00001

shouldn't the condition number be smaller?

In [7]:
# Compute infinity norms
norm_A1_inf = np.linalg.norm(A1, np.inf)
norm_A2_inf = np.linalg.norm(A2, np.inf)
norm_A1_inv_inf = np.linalg.norm(np.linalg.inv(A1), np.inf)
norm_A2_inv_inf = np.linalg.norm(np.linalg.inv(A2), np.inf)

# Compute condition numbers
cond_A1_inf = norm_A1_inf * norm_A1_inv_inf
cond_A2_inf = norm_A2_inf * norm_A2_inv_inf

#print("Infinity norm of A1:", norm_A1_inf)
#print("Infinity norm of A1⁻¹:", norm_A1_inv_inf)
print("Condition number of A1 (∞-norm):", cond_A1_inf)

#print("Infinity norm of A2:", norm_A2_inf)
#print("Infinity norm of A2⁻¹:", norm_A2_inv_inf)
print("Condition number of A2 (∞-norm):", cond_A2_inf)

Condition number of A1 (∞-norm): 9.0
Condition number of A2 (∞-norm): 400004.000007742


![My image](./Screenshot%202025-09-14%20at%2022.51.07.png)


Let's observe a smaller, more observable case:

In [18]:
import numpy as np

# Set print precision
np.set_printoptions(precision=8, suppress=True)

# Symmetric matrix with an almost-zero row/column
A = np.array([
    [4, 2, 2, 0],
    [2, 4, 2, 0],
    [2, 2, 4, 0],
    [0, 0, 0, 0.0000001]
], dtype=float)

# Known solution
x_true = np.array([1, 1, 1, 0.0000001], dtype=float)

# Compute right-hand side
b = A @ x_true

# Print results
print("A =\n", A)
print("x_true = ", x_true)
print("b = ", b)


A =
 [[4.        2.        2.        0.       ]
 [2.        4.        2.        0.       ]
 [2.        2.        4.        0.       ]
 [0.        0.        0.        0.0000001]]
x_true =  [1.        1.        1.        0.0000001]
b =  [8. 8. 8. 0.]


In [19]:
# Solve the system
x = np.linalg.solve(A, b)
print("Original system solution:", x)

# Perturb the last b value slightly
b_perturbed = b.copy()
b_perturbed[3] = 0  # Remove tiny signal
x_perturbed = np.linalg.solve(A, b_perturbed)
print("Perturbed b solution:", x_perturbed)

# Remove last row/col to make 3x3 well-conditioned system
A_reduced = A[:3, :3]
b_reduced = b[:3]
x_reduced = np.linalg.solve(A_reduced, b_reduced)
print("Reduced 3x3 system solution:", x_reduced)

# Condition number before (4x4 system)
cond_full = np.linalg.cond(A, np.inf)
print("Condition number (4x4):", cond_full)

# Condition number after removing problematic row/col (3x3 system)
cond_reduced = np.linalg.cond(A_reduced, np.inf)
print("Condition number (3x3):", cond_reduced)

Original system solution: [1.        1.        1.        0.0000001]
Perturbed b solution: [1. 1. 1. 0.]
Reduced 3x3 system solution: [1. 1. 1.]
Condition number (4x4): 80000000.0
Condition number (3x3): 5.0


Random 3x3 example

In [20]:
import numpy as np

# Ill-conditioned 3x3 matrix
A = np.array([
    [1, 1, 1],
    [1, 1.0001, 1],
    [1, 1, 1.0001]
], dtype=float)

# Original right-hand side
b = np.array([3, 3.0001, 3.0001])

# Solve original system
x = np.linalg.solve(A, b)

# Slight perturbation in b
b_perturbed = np.array([3, 3, 3.0001])

# Solve perturbed system
x_perturbed = np.linalg.solve(A, b_perturbed)

print("Original solution:", x)
print("Perturbed solution:", x_perturbed)
print("Difference:", x_perturbed - x)


Original solution: [1. 1. 1.]
Perturbed solution: [2. 0. 1.]
Difference: [ 1. -1.  0.]


In [21]:
import numpy as np

# Creative ill-conditioned 3x3 matrix (symmetric, some zeros)
A = np.array([
    [10, 9.9999, 0],
    [9.9999, 10, 0.0001],
    [0, 0.0001, 0.000001]
], dtype=float)

# Known solution
x_true = np.array([1, 1, 1], dtype=float)

# Compute right-hand side
b = A @ x_true

# Solve original system
x = np.linalg.solve(A, b)

# Slight perturbation in b (tiny change in last entry)
b_perturbed = np.array([b[0], b[1], b[2] + 1e-7])

# Solve perturbed system
x_perturbed = np.linalg.solve(A, b_perturbed)

print("Original solution:", x)
print("Perturbed solution:", x_perturbed)
print("Difference:", x_perturbed - x)

# Condition number with infinity norm
print("Condition number (∞-norm):", np.linalg.cond(A, np.inf))


Original solution: [1. 1. 1.]
Perturbed solution: [0.9989796  1.00102041 0.99795919]
Difference: [-0.0010204   0.00102041 -0.00204081]
Condition number (∞-norm): 816322.3656802252


In [22]:
import numpy as np

# Matrix as before
A = np.array([
    [10, 9.9999, 0],
    [9.9999, 10, 0.0001],
    [0, 0.0001, 0.000001]
], dtype=float)

# Known solution
x_true = np.array([1, 1, 1], dtype=float)
b = A @ x_true

# Compute SVD of A
U, s, Vt = np.linalg.svd(A)

# Smallest singular value and corresponding singular vector in U
u_min = U[:, -1]

# Perturb b *along* the most sensitive direction (u_min)
epsilon = 1e-5
b_perturbed = b + epsilon * u_min

# Solve original and perturbed systems
x = np.linalg.solve(A, b)
x_perturbed = np.linalg.solve(A, b_perturbed)

print("Original solution:", x)
print("Perturbed solution:", x_perturbed)
print("Difference:", x_perturbed - x)
print("Condition number (∞-norm):", np.linalg.cond(A, np.inf))
print("Smallest singular value:", s[-1])


Original solution: [1. 1. 1.]
Perturbed solution: [0.90882622 1.09117502 0.75234293]
Difference: [-0.09117378  0.09117502 -0.24765707]
Condition number (∞-norm): 816322.3656802252
Smallest singular value: 3.581502739303587e-05
